### PR TITLE
fix(downloads): do not download into browser memory

### LINKED ISSUE #318: File downloads should not use createObjectURL
Requires backend support:

https://github.com/cryostatio/cryostat-web/pull/319#discussion_r724502991
https://github.com/cryostatio/cryostat-web/pull/319#discussion_r724504758

Currently, when downloading template XMLs, report HTMLs, or recording JFRs, the application handles firing the HTTP request with the appropriate Authorization header. The remote asset is then downloaded into memory as a blob, assigned an object data URL, and then that URL passed to the browser's native file save dialog to "download" - despite the fact that we have already downloaded the file from the server and are holding it entirely in-memory, without any progress bar or ability to cancel the download on the user's side.

This download flow should be replaced by a normal download flow handled by the browser, where the application simply points the browser at a URL to download from the server, and the browser continues to display a dialog to the user confirming the choice, showing progress in its Downloads indicator, and handling the actual data transfer and memory/disk management.

To do so will require some backend changes so that Authorization headers do not need to be passed with the GET request to download the assets. A Cryostat user session cookie could be used for this purpose, to authenticate a logged-in Cryostat user for the duration of their login session (ex. generate a token and send it back with `Set-Cookie` on `POST /api/v2/auth`?), or as in the blog posts linked in the above comments, there could be a very short-lived token generated for each specific asset download request, which the client must include as a query parameter to download the actual asset file on a subsequent request.

We can probably use JWTs for either user login tokens or for per-request download tokens - need to do more research here. https://github.com/auth0/java-jwt
https://vertx.io/docs/vertx-auth-jwt/java/

### PR DESCRIPTION
**Pull request by andrewazores**
Fixes #318
Related to https://github.com/cryostatio/cryostat/pull/719

Previously, the file download flow (for templates, reports, and even
binary recording files(!)) would download the remote file into the
browser's memory, then create a local objectURL for the blob, assign
that URL to an anchor element, and prompt the user to "download" the
asset from that objectURL. This means that the user's "Cancel" selection
was somewhat meaningless - by the time they see this, the data has already
been fully transferred to their machine, and all they are cancelling is
to write it to disk. Worse, this means that the time to display the
dialog depends on the time to transfer the data to the user's machine.
For a remote OpenShift cluster sending a large JFR binary file, this
time could become very significant.

This commit improves the download handling by instead relying on the
HTML5 "download" attribute of an anchor element to instruct the browser
to download the asset linked by the href, rather than attempt to open
it, and to use a web-client-defined name for the downloaded file (note
that browsers do not respect this attribute for cross-origin requests,
so some file downloads may be badly named in webpack-dev-server setups).
This vastly decreases the time it takes for the browser's Save File
dialog to appear, and allows the user to cancel the action and prevent
the data from being transferred to begin with.

### CODE REVIEW

**Comment by andrewazores** [comment]
@jan-law this is ready for re-review. This is dependent on a large backend PR as well, but this PR on its own can be reviewed mostly independently. If some backend changes are made during the review they shouldn't change the general concept and download flow too much, just perhaps the API paths or response formats.

**Review by jan-law — COMMENTED** [review]
Not sure if this comment applies to the frontend or backend - when I try to archive an active, running recording from `cryostat:9094` with `Basic` authentication enabled, I see an `AccessDeniedException`
```
Nov 12, 2021 6:06:16 PM io.cryostat.core.log.Logger info
INFO: (10.0.2.100:59474): OPTIONS /api/v1/targets/service%3Ajmx%3Armi%3A%2F%2F%2Fjndi%2Frmi%3A%2F%2Fcryostat%3A9094%2Fjmxrmi/recordings/test 200 0ms
Nov 12, 2021 6:06:16 PM io.cryostat.core.log.Logger error
SEVERE: HTTP 500: java.nio.file.AccessDeniedException: /opt/cryostat.d/recordings.d/ONSXE5TJMNSTU2TNPA5HE3LJHIXS6L3KNZSGSL3SNVUTULZPMNZHS33TORQXIORZGA4TIL3KNV4HE3LJ
io.vertx.ext.web.handler.impl.HttpStatusException: Internal Server Error
Caused by: java.util.concurrent.ExecutionException: java.nio.file.AccessDeniedException: /opt/cryostat.d/recordings.d/ONSXE5TJMNSTU2TNPA5HE3LJHIXS6L3KNZSGSL3SNVUTULZPMNZHS33TORQXIORZGA4TIL3KNV4HE3LJ
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1999)
	at io.cryostat.net.web.http.api.v1.TargetRecordingPatchSave.handle(TargetRecordingPatchSave.java:67)
	at io.cryostat.net.web.http.api.v1.TargetRecordingPatchHandler.handleAuthenticated(TargetRecordingPatchHandler.java:108)
	at io.cryostat.net.web.http.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:90)
	at io.cryostat.net.web.http.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:65)
	at io.vertx.ext.web.impl.BlockingHandlerDecorator.lambda$handle$0(BlockingHandlerDecorator.java:48)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313)
	at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.nio.file.AccessDeniedException: /opt/cryostat.d/recordings.d/ONSXE5TJMNSTU2TNPA5HE3LJHIXS6L3KNZSGSL3SNVUTULZPMNZHS33TORQXIORZGA4TIL3KNV4HE3LJ
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:90)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
	at java.base/sun.nio.fs.UnixFileSystemProvider.createDirectory(UnixFileSystemProvider.java:389)
	at java.base/java.nio.file.Files.createDirectory(Files.java:690)
	at io.cryostat.recordings.RecordingArchiveHelper.writeRecordingToDestination(RecordingArchiveHelper.java:306)
	at io.cryostat.recordings.RecordingArchiveHelper.lambda$saveRecording$0(RecordingArchiveHelper.java:138)
	at io.cryostat.net.TargetConnectionManager.executeConnectedTask(TargetConnectionManager.java:136)
	at io.cryostat.recordings.RecordingArchiveHelper.saveRecording(RecordingArchiveHelper.java:131)
	... 11 more
Nov 12, 2021 6:06:16 PM io.cryostat.core.log.Logger info
INFO: (10.0.2.100:59478): PATCH /api/v1/targets/service%3Ajmx%3Armi%3A%2F%2F%2Fjndi%2Frmi%3A%2F%2Fcryostat%3A9094%2Fjmxrmi/recordings/test 500 40ms
```

**Comment by andrewazores** [comment]
> Not sure if this comment applies to the frontend or backend - when I try to archive an active, running recording from `cryostat:9094` with `Basic` authentication enabled, I see an `AccessDeniedException`
> 
> ```
> Nov 12, 2021 6:06:16 PM io.cryostat.core.log.Logger info
> INFO: (10.0.2.100:59474): OPTIONS /api/v1/targets/service%3Ajmx%3Armi%3A%2F%2F%2Fjndi%2Frmi%3A%2F%2Fcryostat%3A9094%2Fjmxrmi/recordings/test 200 0ms
> Nov 12, 2021 6:06:16 PM io.cryostat.core.log.Logger error
> SEVERE: HTTP 500: java.nio.file.AccessDeniedException: /opt/cryostat.d/recordings.d/ONSXE5TJMNSTU2TNPA5HE3LJHIXS6L3KNZSGSL3SNVUTULZPMNZHS33TORQXIORZGA4TIL3KNV4HE3LJ
> io.vertx.ext.web.handler.impl.HttpStatusException: Internal Server Error
> Caused by: java.util.concurrent.ExecutionException: java.nio.file.AccessDeniedException: /opt/cryostat.d/recordings.d/ONSXE5TJMNSTU2TNPA5HE3LJHIXS6L3KNZSGSL3SNVUTULZPMNZHS33TORQXIORZGA4TIL3KNV4HE3LJ
> 	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
> 	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1999)
> 	at io.cryostat.net.web.http.api.v1.TargetRecordingPatchSave.handle(TargetRecordingPatchSave.java:67)
> 	at io.cryostat.net.web.http.api.v1.TargetRecordingPatchHandler.handleAuthenticated(TargetRecordingPatchHandler.java:108)
> 	at io.cryostat.net.web.http.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:90)
> 	at io.cryostat.net.web.http.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:65)
> 	at io.vertx.ext.web.impl.BlockingHandlerDecorator.lambda$handle$0(BlockingHandlerDecorator.java:48)
> 	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313)
> 	at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
> 	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
> 	at java.base/java.lang.Thread.run(Thread.java:829)
> Caused by: java.nio.file.AccessDeniedException: /opt/cryostat.d/recordings.d/ONSXE5TJMNSTU2TNPA5HE3LJHIXS6L3KNZSGSL3SNVUTULZPMNZHS33TORQXIORZGA4TIL3KNV4HE3LJ
> 	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:90)
> 	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
> 	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
> 	at java.base/sun.nio.fs.UnixFileSystemProvider.createDirectory(UnixFileSystemProvider.java:389)
> 	at java.base/java.nio.file.Files.createDirectory(Files.java:690)
> 	at io.cryostat.recordings.RecordingArchiveHelper.writeRecordingToDestination(RecordingArchiveHelper.java:306)
> 	at io.cryostat.recordings.RecordingArchiveHelper.lambda$saveRecording$0(RecordingArchiveHelper.java:138)
> 	at io.cryostat.net.TargetConnectionManager.executeConnectedTask(TargetConnectionManager.java:136)
> 	at io.cryostat.recordings.RecordingArchiveHelper.saveRecording(RecordingArchiveHelper.java:131)
> 	... 11 more
> Nov 12, 2021 6:06:16 PM io.cryostat.core.log.Logger info
> INFO: (10.0.2.100:59478): PATCH /api/v1/targets/service%3Ajmx%3Armi%3A%2F%2F%2Fjndi%2Frmi%3A%2F%2Fcryostat%3A9094%2Fjmxrmi/recordings/test 500 40ms
> ```

This is a backend problem. Are you using `devserver.sh` or `run.sh`? Whichever way, the Cryostat process is lacking permissions to access (create) the archive directory corresponding to the target you're trying to archive from.

File: src/app/Shared/Services/Api.service.tsx [hunk 1]
```diff
@@ -333,65 +333,62 @@ export class ApiService {
...
-      req().subscribe(resp =>
+    const body = new window.FormData();
+    body.append('resource', recording.downloadUrl.replace('/api/v1', '/api/beta'));
+    this.sendRequest('beta', 'auth/token', {
+      method: 'POST',
+      body,
+    })
+      .pipe(
+        concatMap(resp => resp.json()),
+        map(json => json.data.result.resourceUrl)
```
[the remark this answers is not in the document]
Comment (andrewazores) [reply src/app/Shared/Services/Api.service.tsx]: Good idea.

**Comment by jan-law** [comment]
> Are you using `devserver.sh` or `run.sh`?

I'm using `mvn clean package && CRYOSTAT_AUTH_MANAGER=io.cryostat.net.BasicAuthManager CRYOSTAT_CORS_ORIGIN=http://localhost:9000 CRYOSTAT_DISABLE_SSL=true CRYOSTAT_DISABLE_JMX_AUTH=true sh smoketest.sh` for the backend

**Comment by andrewazores** [comment]
> > Are you using `devserver.sh` or `run.sh`?
> 
> I'm using `mvn clean package && CRYOSTAT_AUTH_MANAGER=io.cryostat.net.BasicAuthManager CRYOSTAT_CORS_ORIGIN=http://localhost:9000 CRYOSTAT_DISABLE_SSL=true CRYOSTAT_DISABLE_JMX_AUTH=true sh smoketest.sh` for the backend

Okay, so in that case the `/opt/cryostat.d/recordings.d` directory it's trying to access is actually on a mounted volume as set in `run.sh`. @hareetd also ran into a similar permissions issue with this podman setup recently, but I'm not sure what the difference is between your system and mine. I'll file a bug against Cryostat for it.

**Comment by andrewazores** [comment]
@jan-law check https://github.com/cryostatio/cryostat/issues/751 , I included a diff there that you can apply to your `run.sh` for now. It will mean that killing and restarting your `smoketest.sh` setup will delete any archives, previously uploaded templates, etc. which would normally be persisted in your working directory's `archives`, `templates`, etc. directories normally, but at least it will properly run and not run into AccessDeniedExceptions.

**Review by jan-law — COMMENTED** [review]
Downloading templates and recordings is working well.

If I click "Download Report" on any active or archived recording, the report HTML is displayed in a new tab and the report is not downloaded. Is this the intended behaviour?

**Comment by andrewazores** [comment]
> Downloading templates and recordings is working well.
> 
> If I click "Download Report" on any active or archived recording, the report HTML is displayed in a new tab and the report is not downloaded. Is this the intended behaviour?

I observed this too and thought it was kind of a nice change - it lets you view the document in "full screen" (not in a miniature iframe anyway). You can still do a typical `Ctrl+s` to save the HTML document to disk if you do want to store it permanently.

I think I can instruct the browser to actually download it rather than viewing it in a new tab by setting the [`Content-Disposition`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition) header in the API response that sends it. In this case the browser would directly download the file and the user could then open it again using the browser, for example by just clicking on the file in the browser's downloads area, and it would again open up in that full frame view. I don't have much of an opinion on it either way, although now that you call the change to attention it seems I should either change the `Content-Disposition` to perform the download, or I should rename the action to something like "View Report..." to match the new behaviour.

**Comment by jan-law** [comment]
> I don't have much of an opinion on it either way, although now that you call the change to attention it seems I should either change the Content-Disposition to perform the download, or I should rename the action to something like "View Report..." to match the new behaviour.

I like the idea of a full screen report with a "View Report" action.

**Review by jan-law — APPROVED** [review]
Looks good!

**Pull request merged**